### PR TITLE
nixos/buildkite-agent: move to v3

### PR DESCRIPTION
**Pull request by flokli**
This contains some of the things #77831 and #68378 (which incorporated things from #59358) aim to fix.

#### Diff to #77831:
 - Above PR makes specifying private and public key optional. However, public keys aren't needed at all to clone private git repos, and with only private keys left, we can simplify the config structure. This also includes the nix store import prevention mechanism suggested by @Infinisil in https://github.com/NixOS/nixpkgs/pull/68378#discussion_r322521825
 - Instead of supporting two somewhat similar options `meta-data` and `tags`, we only support the one supported by version 3.x of the module (as 2.x is dropped here anyways - its latest release has been in January 2018))
 - The bug of setting `cfg.package` having no effect currently is fixed as well.

#### Diff to #68378:
 - The `cfg.extraSetup` command, which would run as root, use deprecated `PermissionsStartOnly=yes and requires fixup in preStart was dropped. We still keep `cfg.shell`, which sets `BUILDKITE_SHELL`.
 - The multiple buildkite agent setup isn't included - I find it somewhat complex and would like to discuss a bit more about usage before adding it.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/continuous-integration/buildkite-agent.nix [hunk 1]
```diff
@@ -74,13 +74,12 @@ in
         '';
       };
 
-      meta-data = mkOption {
-        type = types.str;
-        default = "";
-        example = "queue=default,docker=true,ruby2=true";
+      tags = mkOption {
```
Comment (mrkkrp) [inline nixos/modules/services/continuous-integration/buildkite-agent.nix]: If we're dropping support for 2.x, we should also make the `buildkite-agent` to be 3.x version. Right now it is still 2.x so by default you'd get something that is not working: version 2 wants `meta-data` and not `tags`.

File: nixos/modules/services/continuous-integration/buildkite-agent.nix [hunk 2]
```diff
@@ -204,38 +204,39 @@ in
         environment = config.networking.proxy.envVars // {
           HOME = cfg.dataDir;
           NIX_REMOTE = "daemon";
+          BUILDKITE_SHELL = cfg.shell;
```
Comment (mrkkrp) [inline nixos/modules/services/continuous-integration/buildkite-agent.nix]: Why not to add `shell` configuration option to the configuration file instead?

**Review by mrkkrp — COMMENTED** [review]
LGTM, left a comment.

**Comment by flokli** [comment]
This still needs release notes, will push them soon.

File: nixos/modules/services/continuous-integration/buildkite-agent.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/services/continuous-integration/buildkite-agent.nix]: this was cherry-picked from the older PR, I made it part of the configuration file now.

**Comment by flokli** [comment]
I adressed the comments, and added the missing release notes. PTAL.

File: nixos/modules/services/continuous-integration/buildkite-agent.nix [hunk 3]
```diff
@@ -74,13 +74,12 @@ in
         '';
       };
 
-      meta-data = mkOption {
-        type = types.str;
-        default = "";
-        example = "queue=default,docker=true,ruby2=true";
+      tags = mkOption {
+        type = (types.attrsOf types.str);
```
Comment (zimbatm) [inline nixos/modules/services/continuous-integration/buildkite-agent.nix]:
```suggestion
        type = types.attrsOf types.str;
```

**Review by zimbatm — APPROVED** [review]
LGTM

File: nixos/modules/services/continuous-integration/buildkite-agent.nix (hunk 3 above)
Comment (flokli) [reply nixos/modules/services/continuous-integration/buildkite-agent.nix]: done.

**Pull request merged**